eth/filters/filter_system.go

```diff
@@ -14,172 +14,588 @@
+			if ev.Time.After(f.created) {
+				matchedLogs := filterLogs(convertLogs(e.Logs, false), f.logsCrit.Addresses, f.logsCrit.Topics)
+				logHandler(f, matchedLogs)
+			}
+		}
+	}
+
+	m.allMu.Lock()
+	for _, f := range inactive {
+		delete(m.all, f.ID)
+		glog.Warningf("filter 0x%x uninstalled, queue full\n", f.ID)
+	}
+	m.allMu.Unlock()
+
+	// remove filter for event listening, this must be run in a seperate go routine
+	// since timeout is called from the work loop and sending uninstall requests to the
+	// work loop from "itself" may deadlock when the uninstall channel is full.
+	go func() {
+		for _, f := range inactive {
+			m.uninstall <- f
+		}
+	}()
+}
+
+// timeout uninstalls all filters that have not been used in the last 5 minutes.
+func (m *Manager) timeout() {
+	deadline := time.Now().Add(-5 * time.Minute)
+	var inactive []*filter
+	m.allMu.Lock()
+	for _, f := range m.all {
+		if f.lastUsed.Before(deadline) && f.canTimeout {
+			delete(m.all, f.ID) // filter cannot be used from the external
+			inactive = append(inactive, f)
 		}
 	}
+	m.allMu.Unlock()
+
+	// remove filter for event listening, this must be run in a seperate go routine
+	// since timeout is called from the work loop and sending uninstall requests to the
+	// work loop from "itself" may deadlock when the uninstall channel is full.
+	go func() {
+		for _, f := range inactive {
+			m.uninstall <- f // delete for the internal work loop
+		}
+	}()
+}
+
+// run is the manager work loop.
+// It will receive events and forwards them to installed filters.
+// Inactive filters will be uninstalled.
+func (m *Manager) run() {
+	index := make(filterIndex)
+	timeout := time.NewTicker(30 * time.Second)
+
+	for {
+		select {
+		case f := <-m.install:
+			// lazy load
+			if _, found := index[f.typ]; !found {
+				index[f.typ] = make(map[FilterID]*filter)
+			}
+			index[f.typ][f.ID] = f
+		case f := <-m.uninstall:
+			close(f.hashes)
+			close(f.logs)
+			delete(index[f.typ], f.ID)
+		case ev, ok := <-m.sub.Chan():
+			if !ok {
+				glog.V(logger.Debug).Infoln("filter manager stopped")
+				return
+			}
+			m.process(index, ev)
+		case <-timeout.C:
+			m.timeout()
+		}
+	}
+}
+
+// Stop the filter system.
+func (m *Manager) Stop() {
+	m.sub.Unsubscribe() // end worker loop
+}
+
+// convertLogs is a helper utility that converts vm.Logs to []filter.Log.
+func convertLogs(in vm.Logs, removed bool) []Log {
+	logs := make([]Log, len(in))
+	for i, l := range in {
+		logs[i] = Log{l, false}
+	}
+	return logs
+}
+
+// filterIDGenerator helper utility that generates a (pseudo) random sequence of bytes
+// that are used to generate filter identifiers.
+func filterIDGenerator() *rand.Rand {
+	if seed, err := binary.ReadVarint(bufio.NewReader(crand.Reader)); err == nil {
+		return rand.New(rand.NewSource(seed))
+	}
+	return rand.New(rand.NewSource(int64(time.Now().Nanosecond())))
+}
+
+// newID generates a filter identifier.
+func newID() (FilterID, error) {
```

Comment:
Small nitpick, but please move the FilterID constructor stuff closer to the FilterID type.